# **Introdução**

Em programação, uma função é uma sequência nomeada de instruções que executa uma operação de computador. Ao definir uma função, especifica-se: i) o nome; e a ii) sequência de instruções. Depois disso, é possível chamá-la pelo nome quantas vezes forem necessárias.

## **Chamada de Função**

Já vimos até aqui diversos exemplos de chamadas de funções embutidas, como por exemplo:

In [ ]:
type(42)

int

O nome da função acima é *type*. A expressão entre parênteses é chamada de argumento da função. Para esta função, o resultado é o tipo do argumento. É comum dizer que uma função recebe um argumento e retorna um resultado. O resultado também é chamado de valor de retorno.

## **Funções Matemáticas**

*Python* possui diferentes módulos (ou bibliotecas) que fornecem diversas funções úteis. Um desses módulos é o *math* que fornece várias funções matemáticas prontas e que são úteis para realizar cálculos matemáticos. Para usar qualquer módulo em *Python* é necessário importá-lo antes. Por exemplo, para o módulo *math* é necessário acrescentar o seguinte comando antes de executar suas funções:

In [ ]:
import math

Esta instrução cria um objeto de módulo chamado *math*. Ao se exibir o objeto de módulo, são apresentadas informações sobre ele:



In [ ]:
math

<module 'math' (built-in)>

A importação é necessária apenas uma vez para cada programa. Após o comando de importação do módulo, é possível usar as funções definidas por ele. Para acessá-las é preciso especificar o nome do módulo e o nome da função, separados por um ponto. Este formato é chamado de notação de ponto. A seguir exploraremos várias funções elementares.

### **math.sqrt(x)**

É uma função para calcular a raiz quadrada de do argumento x.

In [ ]:
resultado = math.sqrt(9)
print(resultado)

3.0


### **math.cos(x)**

Função para calcular o cosseno do argumento x, em radianos.

In [ ]:
resultado = math.cos(30)
print(resultado)

0.15425144988758405


### **math.tan(x)**

Função para calcular a tangente do argumento x, em radianos.

In [ ]:
resultado = math.tan(3.1417)
print(resultado)

0.00010734641061924728


### **math.sin(x)**

Função para calcular o seno de do argumento x.

In [ ]:
resultado = math.sin(3.1417)
print(resultado)

-0.00010734641000075741


### **Exercício**

Calcule o fatorial de 10. Dica:  pesquisar na documentação (https://docs.python.org/pt-br/3/library/math.html).

### **Principais Funções da Matemática**

Função | Nome | Exemplo | Resultado
---|---|---|---
Raiz quadrada | sqrt | sqrt(9) | 3
Valor absoluto | abs | abs(-2) | 2
Cosseno | cos | cos(30) | 0.1542514
Tangente | tan | tan(7.3456) | 1.7945721
Seno | sin | sin(pi) | 1.2246e-16
Parte inteira | trunc | trunc(5.2) | 5
Maior inteiro menor que | ceil | ceil(5.2) | 6
Menor inteiro maior que | floor | floor(5.2) | 5
Potência | pow | pow(2,3) | 8
Logarítmo natural | log | log(148.41315910257657) | 5
Logarítmo base 10 | log10 | log10(1000) | 3
Logarítmo base 2 | log2 | log2(256) | 8
Logarítmo qualquer base | log | log(81,3) | 4

Existe uma variante da instrução import que importa nomes de um módulo diretamente.  Por exemplo:

In [ ]:
from math import sqrt
print(sqrt(9))

3.0


### **Constantes**

Além das funções, *Python* possui algumas constantes úteis em diversos cálculos matemáticos fornecidos no módulo math.

#### **pi**

A constante matemática $ $ é utilizada em diversas situações. Uma delas é para se calcular o comprimento de uma circunferência:

In [ ]:
r = 100
C = 2 * math.pi * r
print(C)

628.3185307179587


#### **Número de Euler (ou neperiano)**

Também é muito útil em diversas situações, como em logaritmos naturais por exemplo. Uma outra utilização é para o cálculo de juros compostos, onde o montante após um tempo $t$,$M_t$, a uma taxa de juros composto $r$ é definido como: $M_t=M_i*e^{r*t}$, onde $M_t$ é o montante inicial. O programa a seguir pode ser utilizado para calcular o valor a ser pago considerando um montante
inicial de R$ 1.000,00, com juros composto a uma taxa mensal de 1,5%, durante 12 meses.

In [ ]:
Mi = 1000
r = 0.015
t = 12
Mt = Mi * math.e ** (r*t)
print(Mt)

1197.2173631218102


#### **Resumo das constantes**

Constante | Valor
---|---
pi | 3.141592
e | 2.718281
tau | 6.283185
inf | infinito positivo

# **Como Acrescentar Novas Funções?**

Até agora usamos funções que vêm do *Python*, mas também é possível acrescentar novas funções. Uma definição de função especifica o nome de uma nova função e a sequência de instruções que são executadas quando a função é chamada. Abaixo está um exemplo:

In [ ]:
def print_refrao():
  print('\nSalve, ó terra dos altos coqueiros')
  print('De belezas soberbo estendal')
  print('Nova Roma de bravos guerreiros')
  print('Pernambuco imortal, imortal!\n')

**def** é uma palavra-chave que indica uma definição de função. O nome da função é print_refrao. As regras para os nomes da função são as mesmas que as das variáveis: letras, números e sublinhados são legais, mas o primeiro caractere não pode ser um número.

**Atenção:** não podemos usar palavras-chave como nome de uma função e devemos evitar ter uma variável e uma função com mesmo nome.

Os parênteses vazios depois do nome indicam que esta função não usa argumentos. A primeira linha da definição de função chama-se **cabeçalho**; o resto é chamado de **corpo**. O cabeçalho precisa terminar em dois pontos e o corpo precisa estar endentado. Por convenção, a endentação sempre é de quatro espaços. O corpo pode conter mais de uma instrução.

A definição de uma função cria um objeto de função, que tem o tipo function:

In [ ]:
print(print_refrao)

<function print_refrao at 0x7f65ffb7b3b0>


A sintaxe para chamar a nova função é a mesma que a das funções integradas:

In [ ]:
print_refrao()


Salve, ó terra dos altos coqueiros
De belezas soberbo estendal
Nova Roma de bravos guerreiros
Pernambuco imortal, imortal!



Uma vez que a função tenha sido definida, é possível usá-la dentro de outra função. Por exemplo, para repetir o refrao anterior, podemos escrever uma função print_hino_de_pernambuco:

In [ ]:
def print_hino_de_pernambuco():
  print('Coração do Brasil')
  print('Em teu seio corre o sangue de heróis, rubro veio')
  print('Que há de sempre o valor traduzir')
  print('És a fonte da vida e da história')
  print('Desse povo coberto de glória')
  print('O primeiro, talvez, do porvir')
  print_refrao()
  print('Esses montes, e vales, e rios')
  print('Proclamando o valor de teus brios')
  print('Reproduzem batalhas cruéis')
  print('No presente, és a guarda avançada')
  print('Sentinela indormida e sagrada')
  print('Que defende da pátria os lauréis')
  print_refrao()
  print('...')
  print_refrao()

In [ ]:
print_hino_de_pernambuco()

Coração do Brasil
Em teu seio corre o sangue de heróis, rubro veio
Que há de sempre o valor traduzir
És a fonte da vida e da história
Desse povo coberto de glória
O primeiro, talvez, do porvir

Salve, ó terra dos altos coqueiros
De belezas soberbo estendal
Nova Roma de bravos guerreiros
Pernambuco imortal, imortal!

Esses montes, e vales, e rios
Proclamando o valor de teus brios
Reproduzem batalhas cruéis
No presente, és a guarda avançada
Sentinela indormida e sagrada
Que defende da pátria os lauréis

Salve, ó terra dos altos coqueiros
De belezas soberbo estendal
Nova Roma de bravos guerreiros
Pernambuco imortal, imortal!

...

Salve, ó terra dos altos coqueiros
De belezas soberbo estendal
Nova Roma de bravos guerreiros
Pernambuco imortal, imortal!



**Atenção:** é preciso criar uma função antes de executá-la, ou seja, a definição de função tem que ser executada antes que ela seja chamada.

## **Fluxo de Execução**

Para garantir que uma função seja definida antes do seu primeiro uso, é preciso saber a ordem na qual as instruções serão executadas. A execução sempre inicia na primeira instrução do programa. As instruções são executadas uma após a outra, de cima para baixo.

As definições de função não alteram o fluxo de execução do programa, mas lembre-se de que as instruções dentro da função não são executadas até que a função seja chamada. Uma chamada de função é um desvio no fluxo de execução. Em vez de ir para a instrução seguinte, o fluxo salta para o corpo da função, executa as instruções lá, e então volta para continuar de onde parou.

## **Parâmetros e Argumentos**

Algumas funções que vimos até aqui exigem argumentos. Por exemplo, ao chamar *math.sin*, você usa um número como argumento. Algumas funções exigem mais de um argumento: o *math.pow* exige dois, a base e o expoente. Dentro da função, os argumentos são atribuídos a variáveis chamada **parâmetros**. Abaixo é definida uma função que precisa de um argumento:

In [ ]:
def print_2_vezes(texto):
  print(texto)
  print(texto)

Esta função atribui o argumento a um parâmetro chamado texto. Quando a função é chamada, ela exibe o valor do parâmetro duas vezes.

In [ ]:
print_2_vezes(':-)')

:-)
:-)


In [ ]:
print_2_vezes(86)

86
86


In [ ]:
print_2_vezes(math.cos(math.pi))

-1.0
-1.0


Você também pode usar uma variável como argumento:

In [ ]:
nome = 'Tiago Pessoa Ferreira de Lima'
print_2_vezes(nome)

Tiago Pessoa Ferreira de Lima
Tiago Pessoa Ferreira de Lima


## **As variávis e os parâmetros são locais**



Quando você cria uma variável dentro de uma função, ela é **local**, ou seja, ela só existe dentro da função. Exemplo:

In [ ]:
def somar(operando1, operando2):
  resultado = operando1 + operando2
  print(resultado)

In [ ]:
valor1, valor2 = 2, 3
somar(valor1, valor2)

5


Quando a função somar é encerrada, a variável resultado é destruída. Se tentarmos exibí-la, recebemos uma exceção:

In [ ]:
print(resultado)

-0.00010734641000075741


**Observação:** os parâmetros também são locais.

In [ ]:
print(valor1)

2


In [ ]:
print(valor2)

3


In [ ]:
print(operando1)

NameError: ignored

In [ ]:
print(operando2)

## **Diagrama de Pilha**

Para monitorar quais variáveis podem ser usadas e onde, é uma boa ideia desenhar um diagrama da pilha. Assim como diagramas de estado, os diagramas da pilha mostram o valor de cada variável, mas também mostram a função à qual cada variável pertence.

Cada função é representada por um frame (quadro). Um frame é uma caixa com o nome de uma função junto a ele e os parâmetros e as variáveis da função dentro dele. O diagrama da pilha para o exemplo anterior é exibido na figura abaixo:

...

## **Funções com Resultado e Funções Nulas**

Diversas funções que já usamos, como as funções matemáticas, devolvem resultado. Outras funções, como print, executam uma ação, mas não devolvem um valor. Se você atribuir o resultado a uma variável, recebe um valor especial chamado **None**:

In [ ]:
resultado = print_2_vezes(':P')

In [ ]:
print(resultado)

O valor **None** não é o mesmo que a string 'None'. É um valor especial que tem seu próprio tipo:

In [ ]:
print(type(resultado))

As funções que criamos até aqui são todas nulas, ou seja, não devolvem um valor. Vamos apresentar mais adiante funções com resultados.

# **Por que funções?**

- Criar uma nova função dá a oportunidade de nomear um grupo de instruções, o que deixa o programa mais fácil de ler e depurar.
- Funções podem tornar um programa menor, eliminando o código repetitivo. Depois, se fizer alguma alteração, basta fazê-la num só lugar.
- Funções bem projetadas podem ser úteis para muitos programas. Ou seja, você pode reutilizá-la inclusive em outros programas.

# **Depuração**

Uma das habilidades mais importantes que você vai aprender é a depuração. Leia este [artigo](https://zohaib.me/debugging-in-google-collab-notebook/) e elabore um resumo a respeito do assunto.

# **Glossário**
- Função: uma sequência nomeada de declarações que executa alguma operação. Podem receber argumentos ou produzir um resultado;
- Definição de Função: uma instrução que cria uma função nova, especificando seu nome, parâmetros e as instruções;
- Objeto da Função: um valor é criado por uma definição de função. O nome da função é uma variável que se refere a um objeto de função;
- Cabeçalho: a primeira linha de uma função;
- Corpo: a sequência de instruções dentro de uma função
- Parâmetro: um nome usado dentro de uma função para se referir ao valor passado como argumento;
- Chamada de Função: instrução que executa uma função. É composta pelo nome da função seguido de argumentos entre parênteses;
- Variável Local: variável definida dentro de uma função e por este motivo só pode ser visualizada dentro da função;
- Valor de Retorno: o resultado de uma função;
- Função com Resultado: uma função que devolve um valor;
- Função Nula: uma função que devolve None;
- None: um valor especial apresentado por funções nulas;
- Módulo: um arquivo que contém uma coleção de funções relacionadas e outras definições;
- Instrução de Importação: uma instrução que lê um módulo e cria um objeto de módulo;
- Objeto de Módulo: um valor criado por uma instrução import que oferece acesso aos valores definidos em um módulo;
- Notação de Ponto: sintaxe para chamar funções em outro módulo - o nome do módulo seguido do ponto e o nome da função;
- Fluxo de Execução: a ordem na qual as instruções são executadas.
